### PR TITLE
DDF-3398 Support for RFC822Name in SubjectAlternativeName field of user certificate

### PR DESCRIPTION
**Pull request by troymohl**
#### What does this PR do?
Adds support for the Subject Alternative Name RFC822 Name field to be used for an email address in addition to the emailAddress certificate attribute.

#### Who is reviewing it? 
@bdeining 
@adimka 

#### Select relevant component teams: 
https://github.com/orgs/codice/teams/security

#### Choose 2 committers to review/merge the PR. 
(please choose ONLY two committers from below, delete the rest)
@millerw8 
@stustison

#### How should this be tested? (List steps with links to updated documentation)
Full build with unit tests and integration tests.

#### What are the relevant tickets?
[DDF-3398](https://codice.atlassian.net/browse/DDF-3398)

#### Checklist:
- [ ] Documentation Updated
- [X] Update / Add Unit Tests
- [ ] Update / Add Integration Tests
#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 1]
```diff
@@ -204,10 +208,40 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
...
+                      LOGGER.trace("Adding email from SAN: {}", emailAddress);
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      } catch (Exception e) {
+        LOGGER.debug("Unable to get Subject Alternative Names from certificate", e);
```
Comment (bdeining) [inline platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Any chance we can avoid catching a generic `exception` here?

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 2]
```diff
@@ -204,10 +208,40 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
...
       response.setPrincipal(subjectX500Principal);
       if (response.getAdditionalProperties() == null) {
         response.setAdditionalProperties(new HashMap<>());
       }
+
+      try {
+        if (returnedCredential.getCertificates() != null) {
+          Collection<List<?>> subjectAlternativeNames =
+              returnedCredential.getCertificates()[0].getSubjectAlternativeNames();
+          if (CollectionUtils.isNotEmpty(subjectAlternativeNames)) {
```
Comment (bdeining) [inline platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Any chance of `subjectAlternativeNames` being null?  If so, we should use `isNotBlank` instead.

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 3]
```diff
@@ -204,10 +208,40 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
...
+
       response.setPrincipal(subjectX500Principal);
       if (response.getAdditionalProperties() == null) {
         response.setAdditionalProperties(new HashMap<>());
       }
+
+      try {
+        if (returnedCredential.getCertificates() != null) {
+          Collection<List<?>> subjectAlternativeNames =
+              returnedCredential.getCertificates()[0].getSubjectAlternativeNames();
```
Comment (bdeining) [inline platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Can `returnedCredential.getCertificates()` be empty?

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java (hunk 2 above)
Comment (troymohl) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: I thought isNotEmpty checked !=null and !.empty() on a collection already.

Comment (bdeining) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Whoops, looks like `isNotBlank` also accounts for a whitespace only string.  Do we need to worry about that?

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java (hunk 3 above)
Comment (troymohl) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Removing this check as the line previous already has accessed the certificates at this point.

**Comment by bdeining** [comment]
build now

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 4]
```diff
@@ -204,10 +209,38 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
...
       response.setPrincipal(subjectX500Principal);
       if (response.getAdditionalProperties() == null) {
         response.setAdditionalProperties(new HashMap<>());
       }
+
+      try {
+        Collection<List<?>> subjectAlternativeNames =
+            returnedCredential.getCertificates()[0].getSubjectAlternativeNames();
+        if (CollectionUtils.isNotEmpty(subjectAlternativeNames)) {
+          for (List<?> alternativeNameData : subjectAlternativeNames) {
```
Comment (adimka) [inline platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Could we try and flatten this logic a bit with a stream and filters?

Comment (troymohl) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Updated to use stream and separate method to make this a bit more readable.

**Comment by bdeining** [comment]
build now

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 5]
```diff
@@ -204,10 +210,32 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
       Credential returnedCredential = validator.validate(credential, requestData);
       X500Principal subjectX500Principal =
           returnedCredential.getCertificates()[0].getSubjectX500Principal();
+
       response.setPrincipal(subjectX500Principal);
       if (response.getAdditionalProperties() == null) {
         response.setAdditionalProperties(new HashMap<>());
       }
+
+      try {
```
[the remark this answers is not in the document]
Comment (troymohl) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Fixed.

**Comment by troymohl** [comment]
build now

File: platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java [hunk 6]
```diff
@@ -230,6 +240,44 @@ public TokenValidatorResponse validateToken(TokenValidatorParameters tokenParame
...
+            .stream()
+            .filter(sanData -> sanData.size() > 1)
+            .map(this::getRfc822Name)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .forEach(
+                email ->
+                    response
+                        .getAdditionalProperties()
+                        .put(SubjectUtils.EMAIL_ADDRESS_CLAIM_URI, email));
```
[the remark this answers is not in the document]
Comment (troymohl) [reply platform/security/sts/security-sts-pkivalidator/src/main/java/org/codice/ddf/security/validator/pki/PKITokenValidator.java]: Updated to handle a list of emailaddresses.

File: platform/security/sts/security-sts-certificateclaimshandler/src/main/java/org/codice/ddf/security/claims/certificate/CertificateClaimsHandler.java [hunk 7]
```diff
@@ -88,10 +88,17 @@ private void buildClaim(
     if (value == null) {
       return;
     }
+
     ProcessedClaim c = new ProcessedClaim();
     c.setClaimType(claimType);
     c.setPrincipal(principal);
-    c.addValue(value);
+
+    if (value instanceof List<?>) {
```
Comment (Variadicism) [inline platform/security/sts/security-sts-certificateclaimshandler/src/main/java/org/codice/ddf/security/claims/certificate/CertificateClaimsHandler.java]: :pencil2: The `<?>` is unnecessary.

Comment (troymohl) [reply platform/security/sts/security-sts-certificateclaimshandler/src/main/java/org/codice/ddf/security/claims/certificate/CertificateClaimsHandler.java]: Fixed.

**Comment by troymohl** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by bdeining** [comment]
build now

**Comment by LinkMJB** [comment]
build now

**Comment by LinkMJB** [comment]
build now

**Pull request merged**